porcupine-core/src/Control/Arrow/FoldA.hs

```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE Arrows #-}
+
+-- | This module defines the type 'FoldA', which is a generalization of 'Fold'
+-- from the `foldl` package.
+
+module Control.Arrow.FoldA
+  ( module Control.Foldl
+  , FoldA(..)
+  , Pair(..)
+  , generalizeA
+  , specializeA
+  , premapA
+  , postmapA
+  , prefilterA
+
+  , toTup
+  , fromTup
+  , (&&&&)
+  , (****)
+  ) where
+
+import Prelude hiding (id, (.))
+
+import Control.Arrow
+import Control.Category
+import Control.Foldl
+import Data.Profunctor
+
+
+ret :: (Arrow arr) => t -> arr a t
+ret = arr . const
+{-# INLINE ret #-}
+
+data Pair a b = Pair !a !b
```

Comment:
May I ask why all these strict pairs are so important?

Reply:
I stuck with `foldl` implementation. cargo cult maybe, but in that case too strict is better than no strict enough I believe.